src/main/java/fi/aalto/cs/intellij/common/Resources.java

```diff
@@ -0,0 +1,72 @@
+package fi.aalto.cs.intellij.common;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A class that provides methods to read the resource files of the plugin application.
+ * Normally, this class should be accessed via {@code Resources.SingletonUtility.INSTANCE}.
+ */
+class Resources {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(Resources.class);
+
+  private ResourceProvider resourceProvider;
+
+  Resources(@NotNull ClassLoader classLoader) {
+    this(classLoader::getResourceAsStream);
+  }
+
+  Resources(@NotNull ResourceProvider resourceProvider) {
+    this.resourceProvider = resourceProvider;
+  }
+
+  /**
+   * Returns properties from a properties resource file.
+   * @param resourceName Name of a resource.
+   * @return Contents of the resource in a {@link Properties} object.
+   */
+  @Nullable
+  public Properties getProperties(@NotNull String resourceName) {
+    try (InputStream stream = resourceProvider.getResourceAsStream(resourceName)) {
+      return getProperties(stream);
+    } catch (Exception ex) {
+      LOGGER.error("Could not read a resource: " + resourceName, ex);
+      return null;
+    }
+  }
+
+  @Nullable
+  private Properties getProperties(@Nullable InputStream stream) throws IOException {
+    if (stream == null) {
+      LOGGER.error("Stream is null.");
+      return null;
+    }
+    Properties props = new Properties();
+    props.load(stream);
+    return props;
+  }
+
+  @FunctionalInterface
+  interface ResourceProvider {
+    /**
+     * Opens an input stream to the resource indicated by {@code name}.
+     * @param name Name of the resource.
+     * @return An open {@link InputStream} which corresponds to {@code name} or {@code null} if
+     *         the stream could not be opened.
+     */
+    @Nullable
+    InputStream getResourceAsStream(@NotNull String name);
+  }
+
+  public static class SingletonUtility {
+    public static final Resources INSTANCE = new Resources(Resources.class.getClassLoader());
+
+    private SingletonUtility() { }
+  }
+}
```

Comment:
Could you explain why this is a singleton and not a utility class with some static methods? Not saying it's bad, just don't quite follow the logic here

Reply:
It's perhaps an outdated and unnecessary Java pattern (sometimes called 'Bill Pugh Singleton'). See discussion here:
https://github.com/Aalto-LeTech/intellij-plugin/pull/51#discussion_r369008279
I think we should remove it. :+1:

Reply:
you don't remove things cause teammates don't understand :stuck_out_tongue:

Reply:
Sorry I was unclear, I mean the entire Resources class, not the inner class

Reply:
You are right, it perhaps should not be. (And it is not even in a strict sense singleton now because constructors are not private so there is possibility to instantiate multiple objects). What I was thinking that the only use case is to read resources from the our projects class loader so the object would always be constructed the same way. So it seemed a good idea to store that object in a static field and always use the same object instead of creating always a new object.

But now when I think it, it certainly should not be singleton. If the client code wants to reuse a single instance of it, it is the responsibility of theirs, but not the one of this class. :+1:

Reply:
If understood correctly and it's only used with a specific class loader, then one option would be for `resourceProvider` to be a static private, and the methods using it could be static as well? But your approach sounds good to me :+1: